hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -170,6 +179,14 @@ def __stoppable_test_function(self, data):
                 # need to re-raise it so that it will eventually reach the
                 # correct engine.
                 raise
+        finally:
+            if sys.getrecursionlimit() != depth + limit:
+                # If user code has modified the recursion limit within the test
+                # function, silently resetting it would be bad; but *not* resetting
+                # it would lead to a steady increase with each test case.
+                # We therefore raise an explicit error.
+                raise RuntimeError("Do not modify recursion limit within test")
```

Comment:
I'm kinda loathe to turn these into runtime errors - it's certainly a breaking change to do so, but I think I'd not want to even after a deprecation period. Some code is going to tinker with the recursion limit and that seems fine?

I'm actually not sure that silently resetting it *is* bad. It's essentially just maintaining global state that is being modified by a test function, and generally it's better if test functions don't modify global state at all.